funnel/models/email_address.py

```diff
@@ -0,0 +1,623 @@
+            hashlib.md5(  # NOQA: S303 # skipcq: PTC-W1003 # nosec
+                self.email_lower.encode('utf-8')
+            ).hexdigest()
+            if self.email_lower
+            else None
+        )
+
+    def __str__(self) -> str:
+        """Cast email address into a string."""
+        return self.email or ''
+
+    def __repr__(self) -> str:
+        """Debugging representation of the email address."""
+        return f'EmailAddress({self.email!r})'
+
+    def __init__(self, email: str) -> None:
+        if not isinstance(email, str):
+            raise ValueError("A string email address is required")
+        # Set the hash first so the email column validator passes. Both hash columns
+        # are immutable once set, so there are no content validators for them.
+        self.blake2b160 = email_blake2b160_hash(email.lower())
+        self.email = email
+        self.blake2b160_canonical = email_blake2b160_hash(self.email_canonical)
+
+    def is_available_for(self, actor: Any):
+        """Return True if this EmailAddress is available for the given actor"""
+        for backref_name in self.__exclusive_backrefs__:
+            for related_obj in getattr(self, backref_name):
+                user = getattr(related_obj, related_obj.__email_for__)
+                if user is not None and user != actor:
+                    return False
+        return True
+
+    @delivery_state.transition(None, delivery_state.NORMAL)
+    def mark_sent(self) -> None:
+        """Record fact of an email message being sent to this address."""
+        self.delivery_state_at = db.func.utcnow()
+
+    @delivery_state.transition(None, delivery_state.ACTIVE)
+    def mark_active(self) -> None:
+        """Record fact of recipient activity."""
+        self.delivery_state_at = db.func.utcnow()
+
+    @delivery_state.transition(None, delivery_state.SOFT_FAIL)
+    def mark_soft_fail(self) -> None:
+        """Record fact of a soft fail to this email address."""
+        self.delivery_state_at = db.func.utcnow()
+
+    @delivery_state.transition(None, delivery_state.HARD_FAIL)
+    def mark_hard_fail(self) -> None:
+        """Record fact of a hard fail to this email address."""
+        self.delivery_state_at = db.func.utcnow()
+
+    def refcount(self) -> int:
+        """Returns count of references to this EmailAddress instance"""
+        return sum(
+            len(getattr(self, backref_name)) for backref_name in self.__backrefs__
+        )
+
+    @classmethod
+    def mark_blocked(cls, email: str) -> None:
+        """
+        Mark email address as blocked.
+
+        Looks up all existing instances of EmailAddress with the same canonical
+        representation and amends them to forget the email address and set the
+        :attr:`is_blocked` flag.
+        """
+        for obj in cls.get_canonical(email, is_blocked=False).all():
+            obj.email = None
+            obj._is_blocked = True
+
+    @classmethod
+    def get_filter(
```

Comment:
This one seems like a private method. We'll expect `.get()` to be used from everywhere else. Should prefix this with underscore.

Reply:
It's not. It's used in UserEmail and UserEmailClaim.